feugy/melodie
# feat: secures broadcast with TOTP

### What's in there?

Broadcasting on the web increase the attack surface of an application. 
In order to keep user's data as safe as possible, this PR introduces Time-based One-Time Passwords. 

Rules are:
- without TOTP, one can not list and access music, metadatas and artwork.
- the desktop UI always have valid TOTPs
- the web UI requires fresh TOTP: one has to scan or click the barcode displayed in desktop
- when it does not have fresh TOTP, web UI prompts users so they could provide the value
- the app assets (stylesheets, scripts, html files) are still public

Are included here:

- [x] feat(core): enforces valid TOTP when connecting Websocket
- [x] feat(core): stores TOTP secret in database
- [x] feat(core): passes TOTP secret to desktop UI
- [x] feat(core): do not require TOTP on Websocket reconnections
- [x] feat(core): regularly sends new TOTP to client for reconnections
- [x] feat(ui): introduces TOTP store to generate values from secret
- [x] feat(ui): displays TOTP in broadcast button tooltip, with countdown
- [x] feat(ui): includes valid TOTP in broadcast QRCode
- [x] feat(ui): when running on web, initializes store from TOTP url value
- [x] feat(ui): stores server TOTP in local storage to allow web reconections
- [x] feat(ui): when running on web, prompts user from TOTP
- [x] feat(ui): displays TOTP url in settings so user could use generator apps
- [x] refactor(ui): moved JS code from `index.html` file to `index.js`
- [x] refactor(ui): replaces TailwindCSS with WindiCSS
- [x] refactor(site): replaces TailwindCSS with WindiCSS
- [x] refactor(ui): replaces Rollup with Vite
- [x] refactor(site): replaces Rollup with Vite



### How to test?

1. Open desktop: it should work
2. Enable broadcast, then click or scan the barcode: it opens Mélodie on web, and should work
3. Wait for more than 30'', then refresh the web: it should still work
4. Close desktop, then restart (less than 30''), and refresh on web: it should work
5. Close desktop, wait for 30''+ then restart, and refresh on web: it should prompt for TOTP

### Notes to reviewers

When running as desktop, Mélodie's server injects the TOTP secret into the UI, thanks to the url.
It allows the Broadcast button to display valid, fresh TOTP values, and include it into the barcode.

When running on the Web, Mélodie's UI can grab TOTP from url (scan/click desktop barcode), and will receive frequent updates from server (to allow refreshing the page).

The only time when users are prompt for TOTP is on the web, when they got disconnected for more than 30s.
